internal/rbd/controllerserver.go

```diff
@@ -217,6 +217,37 @@ func validateRequestedVolumeSize(rbdVol, parentVol *rbdVolume, rbdSnap *rbdSnaps
 	return nil
 }
 
+func checkValidCreateVolumeRequest(rbdVol, parentVol *rbdVolume, rbdSnap *rbdSnapshot, cr *util.Credentials) error {
+	err := validateRequestedVolumeSize(rbdVol, parentVol, rbdSnap, cr)
+	if err != nil {
+		return err
+	}
+
+	return checkValidEncryptionRequest(rbdVol, parentVol, rbdSnap)
+}
+
+func checkValidEncryptionRequest(rbdVol, parentVol *rbdVolume, rbdSnap *rbdSnapshot) error {
+	// if snapshot is encrypted the request volume should also be encrypted
+	if rbdSnap != nil {
+		if rbdSnap.isEncrypted() && !rbdVol.isEncrypted() {
+			return status.Error(codes.InvalidArgument, "cannot create unencrypted volume from encrypted snapshot")
+		}
+		if !rbdSnap.isEncrypted() && rbdVol.isEncrypted() {
+			return status.Error(codes.InvalidArgument, "cannot create encrypted volume from unencrypted snapshot")
+		}
```

Comment:
```suggestion
		switch {
		case rbdSnap.isEncrypted() && !rbdVol.isEncrypted():
			return status.Error(codes.InvalidArgument, "cannot create unencrypted volume from encrypted snapshot")

		case !rbdSnap.isEncrypted() && rbdVol.isEncrypted():
			return status.Error(codes.InvalidArgument, "cannot create encrypted volume from unencrypted snapshot")
		}
```

Reply:
I don't have any preference. will wait for others' suggestions.

Reply:
I think I prefer the switch/case, but I do not care *that* much either.

Reply:
Yeah. `switch` will be cleaner I believe.